### DISCUSSION TITLE
gitk (part of git-gui) crashes

### DISCUSSION THREAD

**Discussion by hitman-codehq** [Homebrew/discussions#705] [Casks]
I did some reading on brew.sh to try and find the best place to post this so please redirect me if this is not the place.

Since a relatively recent (about 3 weeks ago) brew upgrade gitk is crashing.  It is crashing as described in this post:

https://stackoverflow.com/questions/65938739/when-i-try-to-use-gitk-it-gives-me-autorelease-pool-page-corrupted

Some googling produces quite a few such crashes but most are very old and this is the only one that seems to relate to this problem with Homebrew.  The suggested solution (rm ~/.config/git/gitk) works well, but once you resize the window the problem is back.

I am not sure whether this problem should be reported to Homebrew (you did the port) or the original authors of gitk (they have their own version).  Any tips are appreciated.

**Comment by SMillerDev** [comment]
Does it appear in their version? Report to them
Does it only appear when installed with Homebrew? Report it here

**Reply by rybak** [reply]
[gitk shipped in Git v2.30.1](https://github.com/git/git/blob/v2.30.1/gitk-git/gitk) seems to work correctly.

Here's a comparison between copy from Git repository and what's installed by Homebrew:

```
$ git log -1
commit 773e25afc41b1b6533fa9ae2cd825d0b4a697fad (HEAD, tag: v2.30.1)
Author: Junio C Hamano <gitster@pobox.com>
Date:   Mon Feb 8 14:05:35 2021 -0800

    Git 2.30.1

    Signed-off-by: Junio C Hamano <gitster@pobox.com>
$ git diff --no-index ./gitk-git/gitk /usr/local/Cellar/git-gui/2.30.1/bin/gitk
diff --git a/./gitk-git/gitk b/usr/local/Cellar/git-gui/2.30.1/bin/gitk
index 23d9dd1fe0..d36f7185d9 100755
--- a/./gitk-git/gitk
+++ b/usr/local/Cellar/git-gui/2.30.1/bin/gitk
@@ -1,6 +1,6 @@
 #!/bin/sh
 # Tcl ignores the next line -*- tcl -*- \
-exec wish "$0" -- "$@"
+exec /usr/local/opt/tcl-tk/bin/wish "$0" -- "$@"

 # Copyright © 2005-2016 Paul Mackerras.  All rights reserved.
 # This program is free software; it may be used, copied, modified
@@ -12658,7 +12658,7 @@ catch {
     wm iconphoto . -default gitlogo gitlogo32
 }
 # wait for the window to become visible
-tkwait visibility .
+if {![winfo viewable .]} {tkwait visibility .}
 set_window_title
 update
 readrefs
 ```

**Reply by chowyuncat** [reply]
2.30.1 also crashes for me, but only after running and exiting it once.
```
λ /usr/local/Cellar/git-gui/2.30.0_1/bin/gitk
objc[5734]: autorelease pool page 0x7fbd5615c000 corrupted
  magic     0x00000000 0x00000000 0x00000000 0x00000000
  should be 0xa1a1a1a1 0x4f545541 0x454c4552 0x21455341
  pthread   0x11c819dc0
  should be 0x11c819dc0

Abort trap: 6
```

```
λ /usr/local/Cellar/git-gui/2.30.1/bin/gitk
(works, exit)

λ /usr/local/Cellar/git-gui/2.30.1/bin/gitk
objc[78699]: autorelease pool page 0x7fcabc156000 corrupted
  magic     0x31e53f80 0x00007fff 0x8c3fe464 0x00000035
  should be 0xa1a1a1a1 0x4f545541 0x454c4552 0x21455341
  pthread   0x0
  should be 0x1169c9dc0

Abort trap: 6
```

**Comment by hitman-codehq** [comment]
I did some more investigation and this seems to be a problem not with gitk but with the version of tcl-tk delivered with Brew.  There is already a version of tcl-tk delivered with Mac OS.  The shell for it can be found in /usr/bin/wish.  If you look in the version of gitk that comes with Brew then it launches Brew's version of tcl-tk like so:

/usr/local/opt/tcl-tk/bin/wish

If you edit Brew's gitk to use the built-in MacOS tcl-tk in /usr/bin/wish then it works without a crash.

I took a look on the official tcl-tk webpage here:

https://www.tcl.tk/software/tcltk/

Unfortunately it appears that there is no official MacOS version of tcl-tk.  I found some information about binary distributions but they all reference 3rd party ports:

https://wiki.tcl-lang.org/page/Apple+Macintosh+and+Tcl%2FTk
https://wiki.tcl-lang.org/page/Binary+Distributions

I guess that Brew is a third party port then.  So that means I cannot compare the Brew build of tcl-tk with the official one, because there isn't one.

Should I raise an issue on Brew for tcl-tk?  It seems the best way forward.

**Reply by carlocab** [reply]
That's a surprising find, as the reason why we try to used Homebrew `tcl-tk` as opposed to the built-in macOS one is that the macOS one is old and full of bugs. (This is also why the Python.org distribution of Python ships their own `tcl-tk` for macOS.) See [this PR](https://github.com/Homebrew/homebrew-core/pull/67378) and the various links in it for more detail.

It might be that this is actually a bug in the newer version (8.6.11) of `tcl-tk` shipped by Homebrew. If you can reproduce the crash with a non-Homebrew `tcl-tk` 8.6.11, then this might be an issue for `tcl-tk`.

If you can't reproduce it with a non-Homebrew `tcl-tk` 8.6.11, then this might be an issue for homebrew/core.

I say "might" here because there's also a possibility that there's an issue with your setup, as other reports of issues with `gitk`/`git-gui` have been resolved (AFAICT).

**Reply by carlocab** [reply]
Can you outline the precise steps you took to get a crash, so I can try to reproduce it?

I tried the following, but I'm not experiencing any problems:
```
brew install git-gui
cd $(brew --repository homebrew/core)
gitk
```
and a GUI opens showing my local homebrew/core repository. Resizing the window works fine.

**Reply by hitman-codehq** [reply]
It happens the second time you open it.  So run gitk in a directory containing a repo, close it and then run it again.  After that, it won't run without crashing until you remove the ~/.config/git/gitk file, at which point it can be executed again.

I did some further investigation by checking out and installing the versions on this page https://github.com/Homebrew/homebrew-core/commits/master/Formula/tcl-tk.rb.

It seems that it was the upgrade from 8.6.10 to 8.6.11 that did it.  SHA 3d72f6b on the 21st of December works and 82ed7b8 on the 7th of January doesn't (the in-between version 514fccb gave me errors when I tried to install it for some reason).

Given that the patch set only changes the versions of the source code that are fetched, the problem definitely lies with the tcl-tk implementation and not brew.  But how does the brew team usually resolve such issues, especially given that there is no easily downloadable MacOS version of tcl-tk?

Many thanks for your help!

**Reply by carlocab** [reply]
> It happens the second time you open it. So run gitk in a directory containing a repo, close it and then run it again. After that, it won't run without crashing until you remove the ~/.config/git/gitk file, at which point it can be executed again.

Nope, I can't reproduce this. I've opened and closed `gitk` several times now, and I'm not getting a crash.

Looking at your [SO post](https://stackoverflow.com/questions/65938739/when-i-try-to-use-gitk-it-gives-me-autorelease-pool-page-corrupted) the issue appears to be in the `geometry(state)` setting. I looked in my `~/.config/git/gitk`, and mine says
```
set geometry(state) normal
```
and it stays that way no matter how many times I open and close `gitk`. To try to reproduce your crash, I changed that line to
```
set geometry(state) zoomed
```
but still no dice. `gitk` ran perfectly fine. The only difference it made was that the `gitk` window covered my entire screen when it opened, instead of covering only a (large) fraction of it.

> Given that the patch set only changes the versions of the source code that are fetched, the problem definitely lies with the tcl-tk implementation and not brew. But how does the brew team usually resolve such issues, especially given that there is no easily downloadable MacOS version of tcl-tk?

If the bug is reproducible, and there is a patch to fix it, then we apply the patch (provided that it's been submitted upstream). The `git-gui` formula carries one such patch at the moment.

Sometimes these patches come from the development branch of the program that just haven't yet made it to the current release. It's possible there's already an upstream fix for this (maybe -- it's still not clear this is an issue in `tcl-tk`, unfortunately) that we can apply as a patch.

**Reply by ap--** [reply]
Hi everyone,

I have the same bug, and can still reproduce it. I currently use the alias suggested in the SO post using sed to replace zoomed with normal before launching gitk.

let me know if there's anything I can do to provide helpful debug output.

Andreas

**Reply by carlocab** [reply]
Since I can't reproduce this, I'm not sure I can be of much help here.

I suggest trying to attach `wish` to a debugger then calling the `gitk` script from there to reproduce the crash and get a backtrace. Depending on the output, you may be able to use this to make a bug report to `tcl-tk` and/or `git`.

**Reply by hitman-codehq** [reply]
Ok, I'll see what I can do.  I'm a bit busy for a few days but will get back to you.  Given that the gitk source is available I might be able to cut out some code and make a reproducible test case that we can give to the tcl-tk people.

It's a bit deeper down the rabbit hole than I wanted to go but I always wanted to know a bit about tcl-tk so here's my chance. :-)

**Reply by jaycverg** [reply]
My temporary workaround that don't need to always reset the config file is to open gitk window not in full screen mode. The crashing happens when it was previously zoomed, manually edit the file to normal, opened gitk again then closing it. The file will be reset to zoomed state.

What I did so far:
1. Gitk was initialy in zoomed state.
2. Edit the config file manually back to normal state.
3. Open gitk.
4. Double clicked the title bar to reset the window state.
5. Resize gitk again to a size large enough for my preference.
6. Close gitk.

So far no crashing for me as long as I always remember to "never maximize that window" ever, lol.

**Reply by lumpidu** [reply]
Same here: removing  ~/.config/git/gitk works for me as well. But doh!

**Reply by disteph** [reply]
I also have the same problem. gitk crashes if `~/.config/git/gitk` contains `set geometry(state) zoomed`, it works if it contains `set geometry(state) normal` instead. And for some reason, whenever gitk actually runs, it will always reset the value to `zoomed` whatever I do (including doing nothing and just closing gitk/wish). I even removed all write permissions to the `~/.config/git/gitk` file to prevent anyone from changing it; but gitk still manages to change it back somehow. What kind of trickery is this? ...that makes me wonder about OSX file permissions...

**Reply by disteph** [reply]
yeah, gitk changes the file permissions back to add u+w to `~/.config/git/gitk`, and then changes the contents. Not cool...

**Reply by johannesluedke** [reply]
yes, it's really weird. My current workaround is 

`alias gitk="sed -i .bak 's/zoomed/normal/g' ~/.config/git/gitk && /usr/local/bin/gitk"` -- I set it back from zoomed to normal everytime before opening it. 

If someone would find an actual fix for that, it would be really nice to have be able to stop using this weird workaround

**Reply by buckyster** [reply]
> ﻿﻿If you edit Brew's gitk to use the built-in MacOS tcl-tk in /usr/bin/wish then it works without a crash.

How do you change Brew's gitk to use /usr/bin/wish ?

**Reply by hitman-codehq** [reply]
Hello @carlocab, perhaps you can help me here.

I have created a small test script that opens a window and zooms it, that crashes with the version of wish that is currently shipped with Homebrew.  I tried compiling TCL/TK on my Mac so that I could do some more investigation but following the instructions on https://www.tcl.tk/doc/howto/compile.html resulted in a version that requires the X windowing system to run and which does not recognise the magic "zoomed" argument that is crashing things.  Not so useful.

How can we go forward with this?  Should I submit an issue directly with the TCL/TK project?  Could you point me to instructions on how to compile the Homebrew version of TCL/TK that does not have an X Window dependency and which contains this zoomed keyword?  I have attached the script, that you can either run directly or run by using "wish hello.tk".

Note that sometimes it doesn't crash, but usually messing around a little bit with what windows are opened usually enables it to be reproduced quite easily.

[hello.zip](https://github.com/Homebrew/discussions/files/6734336/hello.zip)

**Reply by hitman-codehq** [reply]
> > If you edit Brew's gitk to use the built-in MacOS tcl-tk in /usr/bin/wish then it works without a crash.
> 
> How do you change Brew's gitk to use /usr/bin/wish ?

Edit the file /usr/local/bin/gitk and change line 3 from:

exec /usr/local/opt/tcl-tk/bin/wish "$0" -- "$@"

to:

exec /usr/bin/wish "$0" -- "$@"

Note that you'll need to use sudo when running the editor.

**Reply by hitman-codehq** [reply]
Just a quick message to say that with a recent upgrade to MacOS Big Sur, the problem stopped happening for me.  I was trying to reproduce the issue using a self compiled version of TCL/TK but it was quite time consuming so I guess I'll have to be content with the Big Sur "workaround".

**Comment by mark-pictor-csec** [comment]
On Big Sur, and getting a somewhat different crash - if I try to add an "all branches" view, it crashes when I click OK in the dialog box.

Using system wish fixes the problem but I get a deprecation warning that the system version of Tk may be removed in a future release.

If it's helpful...
```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000050
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [14002]

VM Regions Near 0x50:
--> 
    __TEXT                      102fef000-102ff3000    [   16K] r-x/r-x SM=COW  /usr/local/Cellar/tcl-tk/8.6.12/bin/wish8.6

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libtk8.6.dylib                	0x000000010302cd86 PackReqProc + 8
1   libtk8.6.dylib                	0x000000010302c686 ArrangePacking + 335
2   libtcl8.6.dylib               	0x000000010323b2c2 TclServiceIdle + 75
3   libtcl8.6.dylib               	0x000000010321f9b0 Tcl_DoOneEvent + 341
4   libtk8.6.dylib                	0x00000001030cdc7b -[TKContentView(TKWindowEvent) generateExposeEvents:] + 237
5   libtk8.6.dylib                	0x00000001030cd9b5 -[TKContentView(TKWindowEvent) drawRect:] + 152
6   com.apple.AppKit              	0x00007fff22e8646b _NSViewDrawRect + 139
7   com.apple.AppKit              	0x00007fff235b81ef -[NSView _recursive:displayRectIgnoringOpacity:inContext:stopAtLayerBackedViews:] + 2115
...
24  com.apple.HIToolbox           	0x00007fff28754583 _BlockUntilNextEventMatchingListInModeWithFilter + 70
25  com.apple.AppKit              	0x00007fff22d16172 _DPSNextEvent + 864
26  com.apple.AppKit              	0x00007fff22d14945 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1364
27  libtk8.6.dylib                	0x00000001030c957a TkMacOSXEventsSetupProc + 155
28  libtcl8.6.dylib               	0x000000010321f942 Tcl_DoOneEvent + 231
29  libtk8.6.dylib                	0x00000001030cea37 TkMacOSXHandleMapOrUnmap + 97
30  libtk8.6.dylib                	0x00000001030349f2 Tk_UnmapWindow + 126
31  libtk8.6.dylib                	0x000000010302cbf4 PackStructureProc + 84
32  libtk8.6.dylib                	0x00000001030191aa Tk_HandleEvent + 785
33  libtk8.6.dylib                	0x0000000103033f08 Tk_DestroyWindow + 492
34  libtk8.6.dylib                	0x0000000103033e4b Tk_DestroyWindow + 303
35  libtk8.6.dylib                	0x0000000103011f99 Tk_DestroyObjCmd + 99
36  libtcl8.6.dylib               	0x000000010317742d TclNRRunCallbacks + 79
...
```

**Comment by chrstphrchvz** [comment]
This may be resolved in Tk 8.6.13 by https://core.tcl-lang.org/tk/info/d638e7ca3616643e (fix for bug report https://core.tcl-lang.org/tk/info/fc5073a451), however I have not managed to test this myself with gitk.

**Reply by chrstphrchvz** [reply]
I have since tried to explain similar crashes more thoroughly (e.g. https://core.tcl-lang.org/tk/info/bb7f5785a5), and I would say that any `autorelease pool page … corrupted` or `Invalid or prematurely-freed autorelease pool` crash in Tk Aqua shares the same cause and is mostly worked around in Tk 8.6.13.

**Comment by hitman-codehq** [comment]
I had actually forgotten about this bug report because (at least for me) this is now working perfectly again.  I tried with MacOS Monterey on both an x86_64 and M1, and also with the previous Big Sur.  No problems with resizing windows or having to do the delete-the-config-file hack.
